File: leetcode/p0084/Solution.java
/*
https://leetcode.com/problems/largest-rectangle-in-histogram/description/

Given n non-negative integers representing the histogram's bar height where the width of each bar is 1, find the area of largest rectangle in the histogram.

 


Above is a histogram where width of each bar is 1, given height = [2,1,5,6,2,3].

 


The largest rectangle is shown in the shaded area, which has area = 10 unit.

 

Example:

Input: [2,1,5,6,2,3]
Output: 10
*/
class Solution {
    
    private int solve(int[] h, int s, int e) {
        if (s == e) {
            return h[s];
        }
        int m = s + (e-s)/2;
        int max = Math.max(solve(h,s,m),solve(h,m+1,e));
        
        int l = m;
        int r = m+1;
        int height = Math.min(h[l],h[r]);
        max = Math.max(max, height * (r - l + 1));
        
        while (l > s || r < e) {
            if (r >= e || (l > s && h[l-1] > h[r+1])) {
                l--;
                height = Math.min(height,h[l]);
            } else {
                r++;
                height = Math.min(height,h[r]);
            }
            max = Math.max(max,height * (r - l + 1));
        }
        return max;
    }
    
    public int largestRectangleArea(int[] heights) {
        if (heights.length == 0) {
            return 0;
        }
        
        return solve(heights, 0, heights.length -1);
    }
}

File: leetcode/p0084/Solution.kt
class Solution {
    fun largestRectangleArea(heights: IntArray): Int {
        return solve(heights, 0, heights.size - 1)
    }
    
    private fun solve(heights: IntArray, left: Int, right: Int): Int {
        if (left == right) {
            return heights[left]
        }
        
        val mid = left + (right - left)/2
        val leftMax = solve(heights, left, mid)
        val rightMax = solve(heights, mid + 1, right)
        
        var maxArea = heights[mid]
        var minHeight = heights[mid]
        var l = mid
        var r = mid
        while (l > left || r < right) {
            val lh = if (l > left) heights[l - 1] else -1
            val rh = if (r < right) heights[r + 1] else -1
            if (lh > rh) {
                l--
                minHeight = Math.min(minHeight, lh)
            } else {
                r++
                minHeight = Math.min(minHeight, rh)
            }
            maxArea = Math.max(maxArea, (r - l + 1) * minHeight)
        }
        return Math.max(Math.max(leftMax, rightMax), maxArea)
    }
}
